### PR TITLE
 sync with head of NOAA-EMC UPP develop

### LINKED ISSUE #844: update inline post to allow number concentration on pressure levels
## Description
The Water in the West project would like to have number concentration on pressure levels in the inline post output

## Solution

Update the upp hash to include this pull request:

- https://github.com/NOAA-EMC/UPP/pull/974

## Alternatives
AI analysis of raindancing rituals.

## Testing:

Tested in the HAFS-AR variant of the HAFS workflow in a large stationary regional domain over the north-east Pacific. Values looked reasonable.

Regression tests in progress.

## Dependent PRs:

- depends on upp submodule https://github.com/NOAA-EMC/UPP/pull/974

### LINKED ISSUE #846: Inline post does not send 4D biomass burning (EBB) variable to UPP
## Description
The newest version of UPP has a 4D variable (in module VRBLS4D) called "EBB" which replaces its prior 2D variable (in module VRBLS2D). FV3's inline post (io/post_fv3.F90) tries to access the 2D array, which no longer exist. This must be corrected before the UPP hash can update.

It appears EBB(:,:,:,1) comes from the ebu_smoke variable. Only the first index (:,:,:,1) of the array is ever used. This change came from @EricJames-NOAA. Nobody made the corresponding change to the inline post. Indeed, they couldn't since there is a library version incompatibility between UPP and ufs-weather-model (see https://github.com/ufs-community/ufs-weather-model/issues/2322).

I think the solution is to send ebu_smoke into ebb(:,:,:,1) in post_fv3.F90.

### To Reproduce:
What compilers/machines are you seeing this with?
Give explicit steps to reproduce the behavior.
1. Update the upp submodule to the head of UPP's develop branch.
2. Update the ufs-weather-model to use the upp-addon-env Spack Stack.
3. Try to compile with the Intel compiler on Hera. (Spack Stack has been updated there.)
4. Witness the errors in accessing the EBB variable in post_fv3.F90

## Additional context
This problem came up when trying to add new fields to the inline post. To do that, we must update the UPP submodule hash to the head of develop. After I got past the library incompatibility, I ran into this EBB trouble.

## Output

N/A

## Testing:

No code changes yet.

## Dependent PRs:

N/A

### LINKED ISSUE #863: spack.yaml should use explicit library version dependencies instead of "develop" or "upp"
## Description
In `ci/spack.yaml`, there are two problems with the github workflow

1. `upp@develop` - uses the spack repository's upp dependencies instead of those of the upp submodule
2. `ip@develop` - uses the latest ip instead of the one that is correct for this version of FV3

## Solution
Provide explicit library dependency versions for all required libraries.

## Alternatives
Ignore the problem and hope for the best.
Note: That alternative approach has failed.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
## Description

This updates the upp submodule to the head of NOAA-EMC UPP's develop branch.

Also, the spack.yaml has explicit UPP dependencies (g2, g2tmpl, etc.) and all libraries have exact version numbers. These version numbers match the top of ufs-weather-model's develop branch for hercules.

### Issue(s) addressed

- fixes #844 
- fixes #846 
- fixes #863 

#### How were these changes tested?  

Tested in the HAFS-AR variant of the HAFS workflow in a large stationary regional domain over the north-east Pacific. Values looked reasonable. Also ran UPP and ufs-weather-model regression tests.

#### What compilers / HPCs was it tested with?  

Hera intel and Hera gnu.

#### Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  

The ufs-weather-model regression tests already test the inline post. However, I added three new number concentration fields to the ufs-weather-model gnv1_nested regression test.

#### Have the ufs-weather-model regression test been run? On what platform?  

The ufs-weather-model regression testing is in progress now. Baseline generation on Hera succeeded.

UPP's own regression tests passed. That tests the offline post (running UPP outside FV3).

#### Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.

Results of many tests should change due to updating the UPP hash.

#### Please commit the regression test log files in your ufs-weather-model branch

Will do.

## Dependencies

None.

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA When you submit ufs-weather-model PR with upp submodule updating to the latest commit from UPP repository, all UPP control files post-xconfig*.txt (under ufs-weather-model/test/parm) need to be recreated for UFS RTs.

**Comment by SamuelTrahanNOAA** [comment]
> When you submit ufs-weather-model PR with upp submodule updating to the latest commit from UPP repository, all UPP control files post-xconfig*.txt (under ufs-weather-model/test/parm) need to be recreated for UFS RTs.

I need your help with that. I'm uncertain of which txt file in tests/parm corresponds to which file in FV3/upp/parm. Also, I'm wondering if any need modification before copying to tests/parm.

**Comment by SamuelTrahanNOAA** [comment]
These three files from tests/parm don't exist in FV3/upp/parm:

- `postxconfig-NT-fv3lam.txt`
- `postxconfig-NT-gfs.txt`
- `postxconfig-NT-gfs_FH00.txt`

This does exist, but it's the wrong one:
- `postxconfig-NT-hafs.txt`

The tests/parm version has no synthetic satellite brightness temperature fields, while the FV3/upp/parm one does. It isn't the `postxconfig-NT-hafs_nosat.txt` either, since that one differs in other ways.

**Comment by WenMeng-NOAA** [comment]
> These three files from tests/parm don't exist in FV3/upp/parm:
> 
> * `postxconfig-NT-fv3lam.txt`
> * `postxconfig-NT-gfs.txt`
> * `postxconfig-NT-gfs_FH00.txt`
> 
> This does exist, but it's the wrong one:
> 
> * `postxconfig-NT-hafs.txt`
> 
> The tests/parm version has no synthetic satellite brightness temperature fields, while the FV3/upp/parm one does. It isn't the `postxconfig-NT-hafs_nosat.txt` either, since that one differs in other ways.

These are simple versions of UPP control files for inline post testing for gfs, rffs and hafs. I will find corresponding xml files and provide you updated txt files later.

**Comment by SamuelTrahanNOAA** [comment]
@WenMeng-NOAA - As mentioned in [this comment](https://github.com/NOAA-EMC/UPP/pull/974#issuecomment-2166807031), I need help debugging the inline post.


```
234:  get_g2_fixedsurfacetypes key:          255  not found in table 4.5
```

You can find the output here:

LOG: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_3037467/gnv1_nested_intel.log
RUN: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_3037467/gnv1_nested_intel

Many ranks give the same error many times, but it's always the same error.

Code is here:

https://github.com/ufs-community/ufs-weather-model/pull/2326

It only works on Hera because only Hera has g2tmpl 1.12.0 in the same Spack Stack as ufs-weather-model prerequisites.

**Comment by SamuelTrahanNOAA** [comment]
There are a multitude of warnings coming from a library when the latest UPP is used.

`148:  get_g2_fixedsurfacetypes key:          255  not found in table 4.5`

@WenMeng-NOAA has tracked this down to the UPP pull request https://github.com/NOAA-EMC/UPP/pull/929. The warnings vanish if you revert the changes to the postxconfig files (or rather, to the post_avblflds.xml used to generate them). The warnings don't seem to have an effect on the output.

~~This warning probably originates from a bug in one of the grib2 libraries (g2 or g2tmpl) or an incompatibility between them. Library updates can take a *long* time.~~

~~Question to everyone: Should we wait for the warnings to go away before updating UPP in fv3atm? Or should we live with the warnings for now?~~

EDIT: We have a bug fix from the UPP side that eliminates these error messages. It is in this PR.

**Comment by SamuelTrahanNOAA** [comment]
Good news, everyone!

We have a fix for the "255  not found in table 4.5" messages!

That means this PR can continue review without thousands of error messages from g2tmpl.

**Comment by SamuelTrahanNOAA** [comment]
I'd like @DusanJovic-NOAA @WenMeng-NOAA and @EricJames-NOAA to look at this PR.

**Review by EricJames-NOAA — APPROVED** [review]
Looks good to me.  Thanks for working on this.

**Comment by zhanglikate** [comment]
I would suggest the following way to include the 3D rhomid

-        l=lm
         do j=jsta,jend
           do i=ista,iend
-
+            do l=1,lm
             tv = t(i,j,l) * (h1+d608*MAX(q(I,J,L),qmin))
             rhomid(i,j,l) = pmid(i,j,l) / (rd*tv)
+            enddo
 
             dustcb(i,j) = MAX(dustcb(i,j), 0.0)
             dustallcb(i,j) = MAX(dustallcb(i,j), 0.0)
@@ -4608,6 +4608,7 @@ module post_fv3
             pp25cb(i,j) = MAX(pp25cb(i,j), 0.0)
             pp10cb(i,j) = MAX(pp10cb(i,j), 0.0)
 
+           l=lm
            ! Surface PM25 dust and seasalt      
            dustpm(i,j)=(dust(i,j,l,1)+0.38*dust(i,j,l,2))*rhomid(i,j,l) !ug/m3
            dustpm10(i,j)=(dust(i,j,l,1)+dust(i,j,l,2)+dust(i,j,l,3)+ &

**Comment by SamuelTrahanNOAA** [comment]
@zhanglikate - Your message was garbled, so I don't know what you want.

Please do one of these:
1. Use the github "suggested change" feature, or
2. Open a PR to this branch with your changes.

File: io/post_fv3.F90 [hunk 1]
```diff
@@ -4609,12 +4758,18 @@ subroutine set_postvars_fv3(wrt_int_state,grid_id,mype,mpicomp)
           enddo
         enddo ! do loop for l
 
-        l=lm
-        do j=jsta,jend
+        do l=1,lm
```
Comment (zhanglikate) [inline io/post_fv3.F90:4761]: Please use modify this part as the following: 
        do j=jsta,jend
          do i=ista,iend
            do l=1,lm
            tv = max(t(i,j,l) * (h1+d608*MAX(q(I,J,L),1e-8)),1e-8)
            rhomid(i,j,l) = pmid(i,j,l) / (rd*tv)
            enddo

            dustcb(i,j) = MAX(dustcb(i,j), 0.0)
            dustallcb(i,j) = MAX(dustallcb(i,j), 0.0)
            sscb(i,j) = MAX(sscb(i,j), 0.0)
            ssallcb(i,j) = MAX(ssallcb(i,j), 0.0)
            bccb(i,j) = MAX(bccb(i,j), 0.0)
            occb(i,j) = MAX(occb(i,j), 0.0)
            sulfcb(i,j) = MAX(sulfcb(i,j), 0.0)
            if(nasa_on) then
              no3cb(i,j) = MAX(no3cb(i,j), 0.0)
              nh4cb(i,j) = MAX(nh4cb(i,j), 0.0)
            endif
            pp25cb(i,j) = MAX(pp25cb(i,j), 0.0)
            pp10cb(i,j) = MAX(pp10cb(i,j), 0.0)

            l=lm
           ! Surface PM25 dust and seasalt      
           dustpm(i,j)=(dust(i,j,l,1)+0.38*dust(i,j,l,2))*rhomid(i,j,l) !ug/m3
           dustpm10(i,j)=(dust(i,j,l,1)+dust(i,j,l,2)+dust(i,j,l,3)+ &
             0.74*dust(i,j,l,4))*rhomid(i,j,l) !ug/m3
           sspm(i,j)=(salt(i,j,l,1)+salt(i,j,l,2)+ &
             0.83*salt(i,j,l,3))*rhomid(i,j,l)  !ug/m3

**Comment by zhanglikate** [comment]
Hi Sam,

       Where is the "suggested change” feature?  I added comments in the place as “suggested change”. Thanks.

File: io/post_fv3.F90 (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply io/post_fv3.F90:4761]: Those lines are already in the file, in other loops. Look for `! chmical field computation` in the file. (Typo is in the original comment.)

That page has instructions for the "suggested change" feature:

- https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA Could you update the upp harsh to the latest commit @81b38a8 from the UPP repository?

**Comment by SamuelTrahanNOAA** [comment]
Yes. It matches that hash now.

**Comment by SamuelTrahanNOAA** [comment]
Could I get some reviews by people "in the know?"
Maybe these individuals?
- Smoke/Chem - @haiqinli @zhanglikate 
- UPP - @WenMeng-NOAA 
- Write component - @DusanJovic-NOAA @junwang-noaa

**Comment by WenMeng-NOAA** [comment]
The upp submodule update and inline post interface "post_fv3.F90" look good to me.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by DusanJovic-NOAA** [comment]
Github workflow failed:
```
2024-08-14T18:45:15.8799819Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:499:53:
2024-08-14T18:45:15.8800893Z 
2024-08-14T18:45:15.8801377Z   499 |                            g2sec4_temp0,g2sec4_temp8,g2sec4_temp9,g2sec4_temp44,         &
2024-08-14T18:45:15.8802380Z       |                                                     1
2024-08-14T18:45:15.8806022Z Error: Symbol ‘g2sec4_temp9’ referenced at (1) not found in module ‘grib2_all_tables_module’
2024-08-14T18:45:15.8807542Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:499:80:
2024-08-14T18:45:15.8808321Z 
2024-08-14T18:45:15.8808820Z   499 |                            g2sec4_temp0,g2sec4_temp8,g2sec4_temp9,g2sec4_temp44,         &
2024-08-14T18:45:15.8809907Z       |                                                                                1
2024-08-14T18:45:15.8811126Z Error: Symbol ‘g2sec4_temp46’ referenced at (1) not found in module ‘grib2_all_tables_module’
2024-08-14T18:45:15.8812522Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:501:79:
2024-08-14T18:45:15.8813317Z 
2024-08-14T18:45:15.8813853Z   501 |                            g2sec5_temp3,g2sec5_temp40,get_g2_sec5packingmethod,          &
2024-08-14T18:45:15.8814838Z       |                                                                               1
2024-08-14T18:45:15.8816070Z Error: Symbol ‘g2sec4_temp49’ referenced at (1) not found in module ‘grib2_all_tables_module’
2024-08-14T18:45:15.8986194Z make[2]: *** [upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/build.make:868: upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/grib2_module.f.o] Error 1
2024-08-14T18:45:15.8989127Z make[1]: *** [CMakeFiles/Makefile2:1278: upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/all] Error 2
2024-08-14T18:45:15.8990581Z make[1]: *** Waiting for unfinished jobs....
```

**Comment by SamuelTrahanNOAA** [comment]
That looks a lot like an out-of-date library. Maybe @RatkoVasic-NOAA can comment?

**Comment by SamuelTrahanNOAA** [comment]
The github workflow gets its UPP dependencies for spack from this line in `ci/spack.yaml`:

```yaml
- upp@develop
```

The `upp@develop` refers to whatever is in spack develop.

In other words, FV3 is getting its UPP dependencies from the spack repository instead of the UPP repository.

**Comment by WenMeng-NOAA** [comment]
> Github workflow failed:
> 
> ```
> 2024-08-14T18:45:15.8799819Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:499:53:
> 2024-08-14T18:45:15.8800893Z 
> 2024-08-14T18:45:15.8801377Z   499 |                            g2sec4_temp0,g2sec4_temp8,g2sec4_temp9,g2sec4_temp44,         &
> 2024-08-14T18:45:15.8802380Z       |                                                     1
> 2024-08-14T18:45:15.8806022Z Error: Symbol ‘g2sec4_temp9’ referenced at (1) not found in module ‘grib2_all_tables_module’
> 2024-08-14T18:45:15.8807542Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:499:80:
> 2024-08-14T18:45:15.8808321Z 
> 2024-08-14T18:45:15.8808820Z   499 |                            g2sec4_temp0,g2sec4_temp8,g2sec4_temp9,g2sec4_temp44,         &
> 2024-08-14T18:45:15.8809907Z       |                                                                                1
> 2024-08-14T18:45:15.8811126Z Error: Symbol ‘g2sec4_temp46’ referenced at (1) not found in module ‘grib2_all_tables_module’
> 2024-08-14T18:45:15.8812522Z /home/runner/work/fv3atm/fv3atm/fv3atm/upp/sorc/ncep_post.fd/grib2_module.f:501:79:
> 2024-08-14T18:45:15.8813317Z 
> 2024-08-14T18:45:15.8813853Z   501 |                            g2sec5_temp3,g2sec5_temp40,get_g2_sec5packingmethod,          &
> 2024-08-14T18:45:15.8814838Z       |                                                                               1
> 2024-08-14T18:45:15.8816070Z Error: Symbol ‘g2sec4_temp49’ referenced at (1) not found in module ‘grib2_all_tables_module’
> 2024-08-14T18:45:15.8986194Z make[2]: *** [upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/build.make:868: upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/grib2_module.f.o] Error 1
> 2024-08-14T18:45:15.8989127Z make[1]: *** [CMakeFiles/Makefile2:1278: upp/sorc/ncep_post.fd/CMakeFiles/upp.dir/all] Error 2
> 2024-08-14T18:45:15.8990581Z make[1]: *** Waiting for unfinished jobs....
> ```

@DusanJovic-NOAA Please check if g2/3.51 and g2tmpl/1.13.0 are used in CI.

**Comment by SamuelTrahanNOAA** [comment]
It's using a cached version of spack, not the latest build.

There should be an update to `.github/workflows/GCC.yml` to increment the cache version.

**Comment by DusanJovic-NOAA** [comment]
> It's using a cached version of spack, not the latest build.
> 
> There should be an update to `.github/workflows/GCC.yml` to increment the cache version.

Actually I think we should ci/spack.yaml, because the cache key is defined based on hash of spack.yaml. So in order to detect potential changes in any of the libraries we should explicitly list every single library and its version , and not use 'upp@develop' which basically 'hides' the changes in upp dependencies.  upp@develop can mean different things at different time. Basically all dependencies must be versioned. And instead of just saying upp, I prefer to have a explicit list of libraries with explicit versions. 

Try to change spack.yaml in your branch and see if it triggers spack cache rebuild.

**Comment by SamuelTrahanNOAA** [comment]
I've updated spack.yaml, and it is rebuilding. There were library version conflicts, so I chose the latest version of the two.

What worries me is this one:

`ip@develop precision=4,d,8`

Using the `@develop` version means the version may change as the branch is retested. Wouldn't we be better off using a specific numbered version? If so, which one?

**Comment by DusanJovic-NOAA** [comment]
> I've updated spack.yaml, and it is rebuilding. There were library version conflicts, so I chose the latest version of the two.
> 
> What worries me is this one:
> 
> `ip@develop precision=4,d,8`
> 
> Using the `@develop` version means the version may change as the branch is retested. Wouldn't we be better off using a specific numbered version? If so, which one?

Yes. All libraries should be versioned, develop is ambiguous. I would use versions from your ufs-weather-model branch ufs_common.lua from modulefiles directory, with updated versions of g2 and g2tmpl. Also, fv3atm does not need sigio, sfcio, nemsio and wrf-io. Those are required by the upp executable, which we are not building here.

**Comment by SamuelTrahanNOAA** [comment]
> Yes. All libraries should be versioned, develop is ambiguous. I would use versions from your ufs-weather-model branch ufs_common.lua from modulefiles directory, with updated versions of g2 and g2tmpl. Also, fv3atm does not need sigio, sfcio, nemsio and wrf-io. Those are required by the upp executable, which we are not building here.

@DusanJovic-NOAA  - Could you check the spack.yaml? It should match your specifications now.

**Comment by DusanJovic-NOAA** [comment]
> > Yes. All libraries should be versioned, develop is ambiguous. I would use versions from your ufs-weather-model branch ufs_common.lua from modulefiles directory, with updated versions of g2 and g2tmpl. Also, fv3atm does not need sigio, sfcio, nemsio and wrf-io. Those are required by the upp executable, which we are not building here.
> 
> @DusanJovic-NOAA - Could you check the spack.yaml? It should match your specifications now.

The spack.yaml looks good. And I see workflow finished successfully. Thank you for fixing this.

**Comment by SamuelTrahanNOAA** [comment]
I made an issue for the spack.yaml troubles:

- #863 

It'll be automatically closed when this PR is merged.

**Comment by zach1221** [comment]
@jkbk2004 testing is finished on WM PR 2326. Feel free to merge this PR.

**Review by jkbk2004 — APPROVED** [review]

**Pull request merged**